modules/ml/src/main/java/org/apache/ignite/ml/recommendation/util/MatrixFactorizationGradient.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
+ *
+ * Licensed under the GridGain Community Edition License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.gridgain.com/products/software/community-edition/gridgain-community-edition-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.ml.recommendation.util;
+
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.Map;
+import org.apache.ignite.ml.math.primitives.vector.Vector;
+
+/**
+ * Gradient of matrix factorization loss function.
+ *
+ * @param <O> Type of an object of recommendation.
+ * @param <S> Type of a subject of recommendation.
+ */
+public class MatrixFactorizationGradient<O extends Serializable, S extends Serializable> implements Serializable {
```

Comment:
I'm not sure but maybe you move gradient application to vector to MatrixFactorizationGradient class? IMHO it looks more clearly and size of RecommendationTrainer class will be decreased

Reply:
Done.